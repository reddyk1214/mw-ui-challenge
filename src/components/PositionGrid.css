.position-grid {
  padding: 12px 12px 0 12px;
}

.position-grid .flex-header {
  color: rgba(255, 255, 255, 0.6);
  font-size: 10px;
  font-weight: normal;
  display: flex;
  gap: 12px;
  border-bottom: 1px solid rgba(255, 255, 255, 0.05);
  line-height: 28px;
}

.position-grid .flex-row {
  display: flex;
  justify-content: space-between;
  gap: 12px;
  font-size: 12px;
  border-top: 1px solid rgba(0, 0, 0, 0.8);
  border-bottom: 1px solid rgba(255, 255, 255, 0.05);
  line-height: 23px;
}

.position-grid .position-grid-border-top {
  border-top: 1px solid rgba(0, 0, 0, 0.8);
}

.position-grid .flex-cell {
  display: flex;
  align-items: center;
}

.position-grid .ticker-column {
  width: 35px;
  padding-left: 8px;
}

.position-grid .name-column {
  flex-grow: 1;
}

.position-grid .ticker-column-data {
  width: 35px;
  color: rgba(255, 255, 255, 0.8);
  padding-left: 8px;
}

.position-grid .name-column-data {
  flex-grow: 1;
  color: rgba(255, 234, 204, 0.7);
}

.position-grid .exposure-column {
  width: 60px;
  justify-content: flex-end;
}

.position-grid .positive {
  color: rgba(6, 222, 255, 0.8);
}

.position-grid .negative {
  color: rgba(242, 0, 50, 0.8);
}
