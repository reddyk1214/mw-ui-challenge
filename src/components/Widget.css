.Widget {
    width: 100%;
    padding: var(--spacing-lg);
    border: 1px solid var(--color-border);
    background-color: var(--color-background);
    border-radius: var(--border-radius);
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.1);
    color: var(--color-text-primary);
  }
  