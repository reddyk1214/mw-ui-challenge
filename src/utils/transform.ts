import type { Position } from '../types';

export function createTransform(
  searchText: string,
): (positions: Array<Position>) => Array<Position> {
  // Replace with your solution
  return null!;
}
