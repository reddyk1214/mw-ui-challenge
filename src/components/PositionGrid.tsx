import type { Position } from '../types';
import './PositionGrid.css';

export interface PositionGridProps {
  positions: Array<Position>;
}

export function PositionGrid(props: PositionGridProps): JSX.Element | null {
  const { positions } = props;

  return (
    <div className="position-grid" role="grid" aria-label="Position data grid">
      <div role="rowgroup" className="flex-header">
        <div role="columnheader" className="flex-cell ticker-column" id="header-ticker">Ticker</div>
        <div role="columnheader" className="flex-cell name-column" id="header-name">Name</div>
        <div role="columnheader" className="flex-cell exposure-column" id="header-exposure">Exposure</div>
      </div>

      <div role="rowgroup">
        {positions.map((stock) => (
          <div role="row" className="flex-row" key={stock.id}>
            <div
              role="gridcell"
              className="flex-cell ticker-column-data"
              aria-labelledby="header-ticker"
            >
              {stock.ticker}
            </div>
            <div
              role="gridcell"
              className="flex-cell name-column-data"
              aria-labelledby="header-name"
            >
              {stock.name}
            </div>
            <div
              role="gridcell"
              className={`flex-cell exposure-column exposure-column-data ${
                stock.exposure >= 0 ? 'positive' : 'negative'
              }`}
              aria-labelledby="header-exposure"
            >
              {stock.exposure?.toLocaleString()}
            </div>
          </div>
        ))}
      </div>
      <div className="position-grid-border-top"></div>
    </div>
  );
}
