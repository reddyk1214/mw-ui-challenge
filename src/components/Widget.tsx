import { useMemo, useState } from 'react';
import type { Position } from '../types';
import { PositionGrid } from './PositionGrid';
import { SearchInput } from './SearchInput';
import { createTransform } from '../utils/transform';
import './Widget.css';

export interface WidgetProps {
  positions: Array<Position>;
}

export function Widget({ positions }: WidgetProps): JSX.Element | null {
  const [searchText, setSearchText] = useState<string>('');

  const filteredPositions = useMemo(() => {
    const transform = createTransform(searchText);
    return transform(positions);
  }, [searchText, positions]);

  return (
    <div className="Widget">
      <SearchInput searchText={searchText} onSearch={setSearchText} /> 
      <PositionGrid positions={filteredPositions} />
    </div>
  );
}
