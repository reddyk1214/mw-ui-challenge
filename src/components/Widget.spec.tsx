// @vitest-environment jsdom
import { render, fireEvent, cleanup } from '@testing-library/react';
import { expect, test, vi, describe, beforeEach, afterEach } from 'vitest';
import { Widget } from './Widget';
import { PositionGrid } from './PositionGrid';
import { createTransform } from '../utils/transform';
import type { Position } from '../types';

vi.mock('./SearchInput', () => ({
  SearchInput: vi.fn(({ searchText, onSearch }) => (
    <input
      data-testid="mock-search-input"
      type="search"
      value={searchText}
      onChange={(e) => onSearch(e.target.value)}
    />
  ))
}));

vi.mock('./PositionGrid', () => ({
  PositionGrid: vi.fn(() => <div data-testid="mock-position-grid" />)
}));

vi.mock('../utils/transform', () => ({
  createTransform: vi.fn(() => vi.fn((positions: Position[]) => positions))
}));

describe('Widget Component', () => {
  let positions: Position[];

  beforeEach(() => {
    positions = [
      { id: '1', ticker: 'AAPL', name: 'Apple Inc', exposure: 100 },
      { id: '2', ticker: 'AMZN', name: 'Amazon.com Inc', exposure: 200 }
    ];
  });

  afterEach(() => {
    cleanup();
    vi.clearAllMocks();
  });

  test('renders SearchInput and PositionGrid components', () => {
    const { container } = render(<Widget positions={positions} />);
    expect(container).toMatchSnapshot();
  });

  test('calls createTransform with the searchText and positions', () => {
    const mockTransform = vi.fn(() => positions);
    (createTransform as jest.Mock).mockReturnValue(mockTransform);

    const { getByTestId } = render(<Widget positions={positions} />);

    expect(createTransform).toHaveBeenCalledWith('');
    expect(mockTransform).toHaveBeenCalledWith(positions);

    const inputElement = getByTestId('mock-search-input');
    fireEvent.change(inputElement, { target: { value: 'AAPL' } });

    expect(createTransform).toHaveBeenCalledWith('AAPL');
  });

  test('passes filtered positions to PositionGrid', () => {
    const filteredPositions = [
      { id: '1', ticker: 'AAPL', name: 'Apple Inc', exposure: 100 }
    ];

    const mockTransform = vi.fn(() => filteredPositions);
    (createTransform as jest.Mock).mockReturnValue(mockTransform);

    render(<Widget positions={positions} />);

    expect(PositionGrid).toHaveBeenCalledWith(
      expect.objectContaining({
        positions: filteredPositions
      }),
      {}
    );
  });

  test('updates PositionGrid when searchText changes', () => {
    const mockTransform = vi.fn((positions: Position[]) => positions);
    (createTransform as jest.Mock).mockReturnValue(mockTransform);

    const { getByTestId } = render(<Widget positions={positions} />);

    const inputElement = getByTestId('mock-search-input');
    fireEvent.change(inputElement, { target: { value: 'AMZN' } });

    expect(createTransform).toHaveBeenCalledWith('AMZN');
    expect(PositionGrid).toHaveBeenCalledWith(
      expect.objectContaining({
        positions: positions
      }),
      {}
    );
  });
});
