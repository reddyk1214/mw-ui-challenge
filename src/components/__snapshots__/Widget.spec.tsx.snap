// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Widget Component > renders SearchInput and PositionGrid components 1`] = `
<div>
  <div
    class="Widget"
  >
    <input
      data-testid="mock-search-input"
      type="search"
      value=""
    />
    <div
      data-testid="mock-position-grid"
    />
  </div>
</div>
`;
